spring-cloud-starter-stream-processor-header-enricher/src/main/java/org/springframework/cloud/stream/app/headerenricher/processor/HeaderEnricherProcessorConfiguration.java

```diff
@@ -0,0 +1,88 @@
+
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Properties;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.boot.json.JsonParserFactory;
+import org.springframework.cloud.stream.annotation.EnableBinding;
+import org.springframework.cloud.stream.messaging.Processor;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Scope;
+import org.springframework.core.convert.ConversionService;
+import org.springframework.core.convert.support.DefaultConversionService;
+import org.springframework.integration.annotation.Transformer;
+import org.springframework.integration.transformer.HeaderEnricher;
+import org.springframework.integration.transformer.support.ExpressionEvaluatingHeaderValueMessageProcessor;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * A Processor app that adds expression-evaluated headers.
+ *
+ * @author Gary Russell
+ */
+@EnableBinding(Processor.class)
+@EnableConfigurationProperties(HeaderEnricherProcessorProperties.class)
+public class HeaderEnricherProcessorConfiguration {
+
+	@Autowired
+	private HeaderEnricherProcessorProperties properties;
+
+	@Bean
+	@Transformer(inputChannel = Processor.INPUT, outputChannel = Processor.OUTPUT)
+	public HeaderEnricher headerEnricher() throws Exception {
+		Map<String, ExpressionEvaluatingHeaderValueMessageProcessor<?>> headersToAdd = new HashMap<>();
+		if (StringUtils.hasText(this.properties.getHeaders())) {
+			try {
+				Map<String, Object> headers = JsonParserFactory.getJsonParser().parseMap(this.properties.getHeaders());
+				for (Entry<String, Object> entry : headers.entrySet()) {
+					headersToAdd.put(entry.getKey(), processor(entry.getValue()));
+				}
+			}
+			catch (Exception e) {
+				ConversionService conversionService = new DefaultConversionService();
+				if (conversionService.canConvert(String.class, Properties.class)) {
+					Properties props = conversionService.convert(this.properties.getHeaders(), Properties.class);
+					Enumeration<?> enumeration = props.propertyNames();
+					while (enumeration.hasMoreElements()) {
+						String propertyName = (String) enumeration.nextElement();
+						headersToAdd.put(propertyName, processor(props.getProperty(propertyName)));
+					}
+				}
+			}
+		}
+		HeaderEnricher headerEnricher = new HeaderEnricher(headersToAdd);
+		headerEnricher.setDefaultOverwrite(this.properties.isOverwrite());
+		return headerEnricher;
+	}
+
+	@Bean
+	@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE) // Need a new processor for each header
+	public ExpressionEvaluatingHeaderValueMessageProcessor<?> processor(Object expression) {
+		Assert.isInstanceOf(String.class, expression);
```

Comment:
I think some `message` would be useful to let end-user to know what's wrong.
Otherwise remember: https://jira.spring.io/browse/INT-4199?